sdks/v4-sdk/src/hook.ts

```diff
@@ -0,0 +1,66 @@
+import invariant from 'tiny-invariant'
+import { isAddress } from 'ethers/lib/utils'
+
+export type HookPermissions = { [key in HookOptions]: boolean }
+
+export enum HookOptions {
+  AfterRemoveLiquidityReturnsDelta = 'afterRemoveLiquidityReturnsDelta',
+  AfterAddLiquidityReturnsDelta = 'afterAddLiquidityReturnsDelta',
+  AfterSwapReturnsDelta = 'afterSwapReturnsDelta',
+  BeforeSwapReturnsDelta = 'beforeSwapReturnsDelta',
+  AfterDonate = 'afterDonate',
+  BeforeDonate = 'beforeDonate',
+  AfterSwap = 'afterSwap',
+  BeforeSwap = 'beforeSwap',
+  AfterRemoveLiquidity = 'afterRemoveLiquidity',
+  BeforeRemoveLiquidity = 'beforeRemoveLiquidity',
+  AfterAddLiquidity = 'afterAddLiquidity',
+  BeforeAddLiquidity = 'beforeAddLiquidity',
+  AfterInitialize = 'afterInitialize',
+  BeforeInitialize = 'beforeInitialize',
+}
+
+export const hookFlagIndex = {
+  [HookOptions.AfterRemoveLiquidityReturnsDelta]: 0,
+  [HookOptions.AfterAddLiquidityReturnsDelta]: 1,
+  [HookOptions.AfterSwapReturnsDelta]: 2,
+  [HookOptions.BeforeSwapReturnsDelta]: 3,
+  [HookOptions.AfterDonate]: 4,
+  [HookOptions.BeforeDonate]: 5,
+  [HookOptions.AfterSwap]: 6,
+  [HookOptions.BeforeSwap]: 7,
+  [HookOptions.AfterRemoveLiquidity]: 8,
+  [HookOptions.BeforeRemoveLiquidity]: 9,
+  [HookOptions.AfterAddLiquidity]: 10,
+  [HookOptions.BeforeAddLiquidity]: 11,
+  [HookOptions.AfterInitialize]: 12,
+  [HookOptions.BeforeInitialize]: 13,
+}
+
+export class Hook {
+  public static permissions(address: string): HookPermissions {
+    invariant(isAddress(address), 'invalid address')
+    if (/0x/.test(address)) address = address.slice(2)
```

Comment:
nit: i feel like some comments/explains would help make the code easier to follow, like why `address.slice(2)` (easier to follow) and `address.slice(36)` (slightly harder to follow)

Reply:
will do